### PR TITLE
Sort list of document types in topological order (according to docume…

### PR DESCRIPTION
**Pull request by geirst**
…nt references) when producing proton config.

This is to ensure that config is applied in the right order in the search backend.

@bjorncs please review
@toregge FYI

### CODE REVIEW

File: config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @author geirst
+ */
+public class TopologicalDocumentTypeSorter {
+
+    private final Map<String, NewDocumentType> unsortedTypes = new LinkedHashMap<>();
+    private final Map<String, NewDocumentType> sortedTypes = new LinkedHashMap<>();
+
+    private TopologicalDocumentTypeSorter(Collection<NewDocumentType> documentTypes) {
+        documentTypes.stream().forEach(docType -> unsortedTypes.put(docType.getName(), docType));
```
Comment (bjorncs) [inline config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java]: No need to call `.stream()`, `Collection` provides `foreach()`.

File: config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        unsortedTypes.values().stream().forEach(docType -> depthFirstTraverse(docType));
```
Comment (bjorncs) [inline config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java]: Same as above.

File: config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    private void depthFirstTraverse(NewDocumentType docType) {
+        // Note that cycles are not allowed and detected earlier in DocumentGraphValidator.
+        if (sortedTypes.containsKey(docType.getName())) {
+            return;
+        }
+        for (NewDocumentType.Name referenceDocTypeName : docType.getDocumentReferences()) {
+            NewDocumentType referenceDocType = unsortedTypes.get(referenceDocTypeName.getName());
+            assert (referenceDocType != null);
```
Comment (bjorncs) [inline config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java]: Note that Java assert statements are typically only enabled when running unit tests, not in production.

File: config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            depthFirstTraverse(referenceDocType);
+        }
+        sortedTypes.put(docType.getName(), docType);
+    }
+
+    public static List<NewDocumentType> sort(Collection<NewDocumentType> documentTypes) {
+        TopologicalDocumentTypeSorter sorter = new TopologicalDocumentTypeSorter(documentTypes);
+        return sorter.sortedTypes.values().stream().collect(Collectors.toList());
```
Comment (bjorncs) [inline config-model/src/main/java/com/yahoo/vespa/model/content/TopologicalDocumentTypeSorter.java]: `return new ArrayList<>(sortedTypes.values());`

**Review by bjorncs — CHANGES_REQUESTED** [review]
See comments

**Comment by geirst** [comment]
@bjorncs please review again

**Review by bjorncs — APPROVED** [review]

**Pull request merged**